packages/backend-common/src/reading/integration.test.ts

```diff
@@ -0,0 +1,154 @@
+        },
+      ],
+      azure: [
+        {
+          host: 'dev.azure.com',
+          // lasts until 2022-01-27
+          token: 'bhs5cbukiuxrkc3ftuyt5h3eqewtkj37lmf3jx5aoajivq3f5jmq',
+        },
+      ],
+    },
+  }),
+});
+
+function withRetries(count: number, fn: () => Promise<void>) {
+  return async () => {
+    let error;
+    for (let i = 0; i < count; i++) {
+      try {
+        await fn();
+        return;
+      } catch (err) {
+        error = err;
+      }
+    }
+    throw error;
+  };
+}
+
+describe('UrlReaders', () => {
+  it(
+    'should read data from azure',
+    withRetries(3, async () => {
+      const data = await reader.read(
+        'https://dev.azure.com/backstage-verification/test-templates/_git/test-templates?path=%2Ftemplate.yaml',
+      );
+      expect(data.toString()).toContain('test-template-azure');
+
+      const res = await reader.readTree(
+        'https://dev.azure.com/backstage-verification/test-templates/_git/test-templates?path=%2F{{cookiecutter.name}}',
+      );
+      const files = await res.files();
+      expect(files).toEqual([
+        {
+          path: 'catalog-info.yaml',
+          content: expect.any(Function),
+        },
+      ]);
+    }),
+  );
+
+  it(
+    'should read data from gitlab',
+    withRetries(3, async () => {
+      const data = await reader.read(
+        'https://gitlab.com/backstage-verification/test-templates/-/blob/master/template.yaml',
+      );
+      expect(data.toString()).toContain('test-template-gitlab');
+
+      const res = await reader.readTree(
+        'https://gitlab.com/backstage-verification/test-templates/-/tree/master/{{cookiecutter.name}}',
+      );
+      const files = await res.files();
+      expect(files).toEqual([
+        {
+          path: 'catalog-info.yaml',
+          content: expect.any(Function),
+        },
+      ]);
+    }),
+  );
+
+  it(
+    'should read data from bitbucket',
+    withRetries(3, async () => {
+      const data = await reader.read(
+        'https://bitbucket.org/backstage-verification/test-template/src/master/template.yaml',
+      );
+      expect(data.toString()).toContain('test-template-bitbucket');
+
+      const res = await reader.readTree(
+        'https://bitbucket.org/backstage-verification/test-template/src/master/{{cookiecutter.name}}',
+      );
+      const files = await res.files();
+      expect(files).toEqual([
+        {
+          path: 'catalog-info.yaml',
+          content: expect.any(Function),
+        },
+      ]);
+    }),
+  );
+
+  it(
+    'should read data from github',
+    withRetries(3, async () => {
+      const data = await reader.read(
+        'https://github.com/backstage-verification/test-templates/blob/master/template.yaml',
+      );
+      expect(data.toString()).toContain('test-template-github');
+
+      const res = await reader.readTree(
+        'https://github.com/backstage-verification/test-templates/tree/master/{{cookiecutter.name}}',
```

Comment:
Good first issue: Add test which works without the branch name? :)

Reply:
Yeah we can build these out to add some important use-cases, but should be a bit careful with adding too many requests so we don't end up hitting ratelimits